File: packages/shared/src/dimensions.ts
import { z } from "zod";

/**
 * Bemassungs-Anker für die 3D-Vorschau – spiegelt
 * services/cad/app/dimensions.py:DimensionSpec. Die Anker kommen vom
 * CAD-Service (POST /dimensions): die gemessene Strecke p1→p2 liegt auf der
 * Geometrie (Template-Koordinaten, effektive/toleranzierte Masse), offset_dir
 * ist die Richtung, in der die Masslinie nach aussen versetzt wird.
 */

export const Vec3 = z.tuple([z.number(), z.number(), z.number()]);
export type Vec3 = z.infer<typeof Vec3>;

export const DimensionSpec = z.object({
  param: z.string(),
  kind: z.enum(["linear", "diameter"]),
  p1: Vec3,
  p2: Vec3,
  offset_dir: Vec3,
});
export type DimensionSpec = z.infer<typeof DimensionSpec>;

/** Antwort von POST /api/cad/dimensions. */
export const DimensionsResponse = z.object({
  archetype: z.string(),
  dims: z.array(DimensionSpec),
});
export type DimensionsResponse = z.infer<typeof DimensionsResponse>;

File: services/cad/app/dimensions.py
"""Bemassungs-Anker für die 3D-Vorschau (Phase 2 der Modell-Bemassung).

Jedes Template kann eine `dimensions(params, profile)`-Funktion registrieren,
die pro editierbarem Parameter einen Anker liefert: die gemessene Strecke
(p1→p2, direkt auf der Geometrie) plus die Richtung, in der die Masslinie
nach aussen versetzt wird. Das Frontend zeichnet daraus Masslinie, Hilfs-
linien und ein Eingabefeld am Modell.

Die Anker rechnen mit denselben effektiven (toleranzierten) Massen wie
build(), damit sie exakt auf den Mesh-Flächen sitzen; angezeigt und editiert
wird im Frontend immer das Nennmass. Damit bleibt die Geometrie-Hoheit beim
Template (Eiserne Regel 1) – das Frontend kennt keine Bauteil-Mathematik.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel

Vec3 = tuple[float, float, float]


class DimensionSpec(BaseModel):
    """Ein bemassbarer Parameter am 3D-Modell (Koordinaten des Templates)."""

    param: str
    kind: Literal["linear", "diameter"] = "linear"
    p1: Vec3
    p2: Vec3
    offset_dir: Vec3
